docs(1367): tidy up for tutorials (#1401)
MetaMask/metamask-docs: snaps/learn/tutorials/gas-estimation.md

---
description: Create a Snap that estimates gas fees.
sidebar_position: 1

> toc_max_heading_level: 4

Author: ```suggestion
```

[`@metamask/create-snap`](https://github.com/MetaMask/snaps/tree/main/packages/create-snap)
starter kit by running:

<Tabs>

>   <TabItem value="yarn" label="yarn" default>

Reviewer: ```suggestion
  <TabItem value="Yarn" label="Yarn" default>
```

Yarn is typically capitalized

yarn create @metamask/snap gas-estimation-snap
```

Next, `cd` into the `gas-estimation-snap` project directory and run:

```bash
yarn install
```

This initializes your development environment with the required dependencies. 

<details>
  <summary>Did you get a warning?</summary>
  <div>
You may get a warning such as: 

```bash
@lavamoat/allow-scripts has detected dependencies without configuration. explicit configuration required.
run "allow-scripts auto" to automatically populate the configuration.
```
You can resolve this error by running:
    
```bash
yarn run allow-scripts auto
```
 </div>
</details>

> ### 2. (Optional) customize your Snap

Reviewer: ```suggestion
### 2. (Optional) Customize your Snap
```

> This Snap is generated from a boilerplate typescript example Snap. While the following steps in the 
> drop-down are optional for testing, they significantly improve the UX by aligning the Snap with its function.
> 
> <details>
>   <summary>Customize the Snap's UX</summary>
>   <div>

Reviewer: ```suggestion
This Snap is generated from a boilerplate TypeScript example Snap.
We recommend customizing your Snap to improve its UX, but this is optional for testing.
If you don't wish to customize your Snap, skip to [Step 3](#3-enable-network-access).

```

In my opinion, we should take the optional sub-steps out of the dropdown and make them into H4 headings (as Joan suggested) so they are visible in the right-hand TOC.

Having such a large amount of content in a dropdown may become cumbersome for users, and looks overly complex. And if we recommend doing the steps anyway, we should keep them visible.

My suggestion also adds a reference to Step 3 if the user wants to skip.

>     2.1 Update your Snap to [display an icon](../best-practices/design-guidelines.md#optimize-your-metadata) in MetaMask.

Reviewer: ```suggestion
#### 2.1. Display an icon

Update your Snap to [display an icon](../best-practices/design-guidelines.md#optimize-your-metadata) in MetaMask.
```

Also, add `toc_max_heading_level: 4` to the page metadata to make these headings visible in the TOC.

Author: Hmm -- so I do prefer the headings for the steps, but worried that adding them to toc makes the tutorials look more imposing, was trying to give a more welcoming flow. For sure, the div has to go if we toc4 cause the links don't work into divs. Will do that now, come back to me on toc4 decision.

>     2.1 Update your Snap to [display an icon](../best-practices/design-guidelines.md#optimize-your-metadata) in MetaMask.

Author: The reason I put it as a div it to put it away --> I am going to go do that for any optional steps in Wallet docs also. This addresses the issue that the tutorial exists for 2 purposes:
1. A quick n dirty "Hello World"
2. A training exercise to awaken awareness to best practice, such as icon etc for good UX

>     2.1 Update your Snap to [display an icon](../best-practices/design-guidelines.md#optimize-your-metadata) in MetaMask.

Author: oh, sorry do you mean the 2.1 ***, 2.2 *** ? It might work, can't use heading at the <summary>, but within div ... checking

>     2.1 Update your Snap to [display an icon](../best-practices/design-guidelines.md#optimize-your-metadata) in MetaMask.

Author: yes, works --> applying

    ├─ packages/
    │  ├─ site/
    |  |  |- src/
    |  |  |  |- App.tsx
    |  |  ├─ package.json
    |  |  |- ...(react app content)
    |  |
    │  ├─ snap/
    |  |  ├─ images/
    |  |  |  |- gas.svg
    |  |  ├─ src/
    |  |  |  |- index.test.ts
    |  |  |  |- index.ts
    |  |  ├─ snap.manifest.json
    |  |  ├─ package.json
    |  |  |- ... (Snap content)
    ├─ package.json
    ├─ ... (other stuff)
    ```

    Open `packages/snap/snap.manifest.json` in a text editor. This file contains the main configuration 
    details for your Snap. Edit the `npm` object, within the `location` object, and add `iconPath` with 
    the value `"images/gas.svg"` to point to your new icon:

    ```json title="snap.manifest.json"
    "location": {
      "npm": {
        "filePath": "dist/bundle.js",
        "iconPath": "images/gas.svg",
        "packageName": "snap",
        "registry": "https://registry.npmjs.org/"
      }
    }
    ```

    Open `packages/snap/package.json` in a text editor. Edit the `files` array and reference the 
    `images/` folder:

    ```json title="package.json"
    "files": [
      "dist/",
      "images/",
      "snap.manifest.json"
    ],
    ```

>     2.2 Update your Snap's wallet prompt

Reviewer 2: ```suggestion
    #### 2.2 Update your Snap's wallet prompt
```

Author: applied across both tutorials

>     2.2 Update your Snap's wallet prompt

Reviewer: ```suggestion
#### 2.2. Update your Snap's wallet prompt
```

>     It's important that user's understand what they're agreeing to when responding to wallet prompts. The 
>     prompt uses the `proposedName` of the Snap, currently "Typescript Example" in our boilerplate.

Reviewer: ```suggestion
    It's important that users understand what they're agreeing to when responding to wallet prompts. The 
    prompt uses the `proposedName` of the Snap, currently "Typescript Example" in the boilerplate.
```

I also wonder about the use of "boilerplate," and if "template" would be a more accessible term.

    Open `packages/snap/snap.manifest.json` in a text editor.
    Edit the `"proposedName"` property within the metadata to provide a functional name such as 
    "Gas Estimate Snap":

    ```json title="snap.manifest.json"
    {
      "version": "0.1.0",
      "description": "An example Snap written in TypeScript.",
      "proposedName": "Gas Estimate Snap",
      "repository": {
        "type": "git",
        "url": "https://github.com/MetaMask/template-snap-monorepo.git"

>       },
>       ```

Reviewer: ````suggestion
    },
    ```
````

>     2.3 Update the Snap's button

Reviewer: ```suggestion
#### 2.3. Update your Snap's button
```

For consistency.

    Open `packages/site/src/components/Buttons.tsx` in a text editor.

>     Edit the Button property to provide a functional name such as "Estimate Gas ":

Reviewer: ```suggestion
    Edit the `Button` property to provide a functional name such as "Estimate Gas":
```

  });
```

### 5. Build and test your Snap

Complete the following steps to build and test your Snap:

> #### 5.1. Increment version

Reviewer: On second thought, should this fall under "Next steps" instead of "Build and test your Snap"? Incrementing the version would be a pre-publish step, not really part of building/testing a Snap.

Author: agree

> To build and test your Snap:

Reviewer: ```suggestion
Complete the following steps to build and test your Snap:
```

Also, can we change Step 5 to "Build and test _your_ Snap" for consistency?

Author: yes, done in both

> 5.1 (Optional for testing) open `package.json` in the root directory of the project, and increment 
>   the `"version"` (if the `"version"` is `0.1.0`, increase it to `0.2.0`).

Reviewer: ```suggestion
#### 5.1. (Optional) Increment the version

Open `package.json` in the root directory of the project, and increment 
  the `"version"` (if the `"version"` is `0.1.0`, increase it to `0.2.0`).
  This is optional for testing.
```

If we want to remove the ordered list and use sub-steps like "5.1," I suggest adding these as H4s for visibility and consistency.

Author: yes, I like it better, but not convinced about H4s going into toc --> makes the tutorial seem more imposing

> 5.2 From the command line, run `yarn start`.

Reviewer: ```suggestion
#### 5.2. Build your Snap

From the command line, run `yarn start`.
```

     http://localhost:8000/
   ```

> 5.3 Open [`localhost:8000`](http://localhost:8000/) in your browser (with MetaMask Flask installed).

Reviewer: ```suggestion
Open [`localhost:8000`](http://localhost:8000/) in your browser (with MetaMask Flask installed).
```

This would fall under "5.2. Build your Snap."

   A page like the following displays:

   <img src={require('../../assets/template-snap.png').default} alt="Test dapp with template Snap" style={{border: '1px solid #DCDCDC'}} />

   This is a boilerplate test dapp for installing and testing your Snap.

> 5.4 Select **Connect** to connect Flask to the dapp.

Reviewer: ```suggestion
#### 5.3. Test your Snap

Select **Connect** to connect Flask to the dapp.
```

   After connecting, you're prompted to install the Snap with the following permissions:

   - **Allow websites to communicate directly with this Snap.**
   - **Access the internet.**
   - **Display dialog windows in MetaMask.**

> 5.5 Select **Confirm** > **OK**.

Reviewer: ```suggestion
Next, select **Confirm** > **OK**.
```

This falls under "5.3. Test your Snap."

> 5.6 After installing, the **Send message** button (or **Estimate gas** button, if you followed Step 2) 

Reviewer: ```suggestion
After installing, the **Send message** button (or **Estimate gas** button, if you followed Step 2) 
```

This falls under "5.3. Test your Snap."

  is enabled. Select this button. A dialog prompt displays with the response from the gas fee API:

<p align="center">
<img src={require('../../assets/gas-estimation.png').default} alt="Gas estimation dialog" width="400px" style={{border: '1px solid #DCDCDC'}} />
</p>

Congratulations, you have integrated a public API into MetaMask and displayed real-time gas fee estimates.

> ### What next?

Reviewer: ```suggestion
### Next steps
```

For consistency with tutorials and guides across our Wallet and Services docs, we should keep the "Next steps" heading here.

> Consider:

Reviewer: ```suggestion
To continue refining your Snap, you can complete these next steps:
```

Suggestion, can play around with wording. This is not something we do consistently in our docs, but where possible, we should avoid using fragments in introductory text for lists (for localization purposes): https://learn.microsoft.com/en-us/style-guide/scannable-content/lists#introductory-text

- Improving the Snap's UX:
  - Complete [Step 2](#2-optional-customize-your-snap).

>   - Update `description`: the description of your Snap.

Reviewer: ```suggestion
  - Update `description`: the description of your Snap.
```

Does this affect the Snap's UI, or is it just metadata?

Author: meta - moved

  - Parse the JSON response from the remote API.
  - Format the fees for better readability.

> - Updating the `location` to its published location before publishing your Snap to `npm`, 

Reviewer: ```suggestion
- Updating the `location` in `snap.manifest.json` to its published location before publishing your Snap to `npm`, 
```

#### 5.3. Test your Snap

Select **Connect** to connect Flask to the dapp.
   After connecting, you're prompted to install the Snap with the following permissions:

   - **Allow websites to communicate directly with this Snap.**
   - **Access the internet.**
   - **Display dialog windows in MetaMask.**

Next, select **Confirm** > **OK**.

After installing, the **Send message** button (or **Estimate gas** button, if you followed Step 2) 
is enabled. Select this button. A dialog prompt displays with the response from the gas fee API:

<p align="center">
<img src={require('../../assets/gas-estimation.png').default} alt="Gas estimation dialog" width="400px" style={{border: '1px solid #DCDCDC'}} />
</p>

Congratulations, you have integrated a public API into MetaMask and displayed real-time gas fee estimates.

### Next steps

You can improve your Snap's UX by:
- Completing [Step 2](#2-optional-customize-your-snaps-ux).
- Parsing the JSON response from the remote API.
- Formatting the fees for better readability.

Before publishing a Snap, it's also important to customize the metadata and properties of your app, 
for example:

> - Update the `location` in `snap.manifest.json` to its published location before publishing your 
> Snap to `npm`. 

Reviewer: ```suggestion
- Update the `location` in `snap.manifest.json` to your Snap's published location.
- Update the `description` in `snap.manifest.json` to a description of your Snap.
```
